odoo/addons/test_lint/tests/_odoo_checker_gettext.py

```diff
@@ -0,0 +1,75 @@
+import os
+
+import astroid
+from pylint import checkers, interfaces
+
+
+class OdooBaseChecker(checkers.BaseChecker):
+    __implements__ = interfaces.IAstroidChecker
+    name = 'odoo'
+
+    msgs = {
+        'E8502': (
+            'Bad usage of _ method.',
+            'gettext-variable',
+            'See https://www.odoo.com/documentation/13.0/reference/translations.html#variables'
+        )
+    }
+
+    def _get_func_name(self, node):
+        func_name = (
+            isinstance(node, astroid.Name) and node.name or
+            isinstance(node, astroid.Attribute) and node.attrname or '')
+        return func_name
+
+    def _check_concatenation(self, node):
+        is_bin_op = False
+        if isinstance(node, astroid.BinOp) and node.op in ('%', '+'):
+            # execute("..." % self._table)
+            if (isinstance(node.right, astroid.Attribute) and
+                    not node.right.attrname.startswith('_')):
+                is_bin_op = True
+            if isinstance(node.right, astroid.Tuple):
+                for elt in node.right.elts:
+                    if (isinstance(elt, astroid.Call) and
+                            # Ignoring for this case: execute("..." % _foo(...))
+                            not self._get_func_name(elt.func).startswith('_')):
+                        is_bin_op = True
+
+        is_format = False
+        # execute("...".format(self._table, table=self._table))
+        if (isinstance(node, astroid.Call) and
+                self._get_func_name(node.func) == 'format'):
+            for keyword in node.keywords or []:
+                if (isinstance(keyword.value, astroid.Attribute) and
+                        not keyword.value.attrname.startswith('_')):
+                    is_format = True
+                    break
+            for argument in node.args or []:
+                if (isinstance(argument, astroid.Name) and not argument.name.startswith('_')):
+                    is_format = True
+                    break
+
+        return is_bin_op or is_format
+
+    def _check_variable_inside_gettext(self, node):
+        if not (
+            isinstance(node, astroid.Call) and node.args and
+            isinstance(node.func, astroid.nodes.Name) and
+            node.func.name in ('_', '_lt')
+        ):
+            return False
+        first_arg = node.args[0]
+        is_concatenation = self._check_concatenation(first_arg)
```

Comment:
`_check_concatenation()` makes sense for `execute`, because identifiers must be injected in queries.
But for `gettext` anything else than a string literal is wrong, because the term extraction will not work, will it?
```python
return not(isinstance(first_arg, astroid.Const) and isinstance(first_arg.value, str))
```

Reply:
With the intention of the concatation check (not functional yet) was for borderlines scenarios like 
https://github.com/odoo/odoo/blob/5a30919826bd4072bf3e0e2915b9df271d58ee78/addons/l10n_fr_pos_cert/models/account_fiscal_position.py#L14-L17

do we accept this? (`first_arg` is of type `astroid.node_classes.BinOp`)
This one could be valid (even if unnecessary) but
```py
_(
  "You cannot modify a fiscal position used in a POS order. "
  + "You should archive it and create a new one." + vals
)
```
is certainly not.
Boh fails at the moment.

Reply:
The first example comes from Python developers that don't know the language, as you can simply write
```python
_( 
     "You cannot modify a fiscal position used in a POS order. "
     "You should archive it and create a new one." 
 ) 
```